src/viur/core/modules/file.py

```diff
@@ -869,6 +944,106 @@ def download(self, blobKey: str, fileName: str = "", download: bool = False, sig
+        options: str = "",
+        download: bool = False,
+    ):
+        """
+        Requests an image using the serving url to bypass direct Google requests.
+
+        :param key: a string with one __ seperator. The first part is the host prefix, the second part is the key
+        :param size: the target image size, take a look at the VALID_SIZES
+        :param filename: a random string with an extention, valid extentions are jpg,png,webp
+        :param options: - seperated options:
+            c - crop
+            p - face crop
+            fv - vertrical flip
+            fh - horizontal flip
+            rXXX - rotate 90, 180, 270
+            nu - no upscale
+        :param download: if value = 1 force download, else not
+        :return:
+        """
+
+        VALID_OPTIONS = {
+            "c",
+            "p",
+            "fv",
+            "fh",
+            "r90",
+            "r180",
+            "r270",
+            "nu"
+        }
+
+        VALID_SIZES = {
+            32, 48, 64, 72, 80, 90, 94,
+            104, 110, 120, 128, 144, 150, 160,
+            200, 220, 288,
+            320,
+            400,
+            512, 576,
+            640,
+            720,
+            800,
+            912,
+            1024, 1152, 1280, 1440, 1600
+        }
+
+        VALID_FMTS = {
+            "jpg": "rj",
+            "jpeg": "rj",
+            "png": "rp",
+            "webp": "rw",
+        }
+
+        try:
+            host, value = key.split("__", 1)
+        except ValueError:
+            raise errors.BadRequest("Invalid key provided for serving url")
+
+        if any(c not in conf.search_valid_chars for c in host):
+            raise errors.BadRequest("key contains invalid characters")
+
+        # extract format from filename
+        file_fmt = "webp"
+
+        if filename:
+            fmt = filename.rsplit(".", 1)[-1].lower()
+            if fmt in VALID_FMTS:
+                file_fmt = fmt
+
+        url = f"https://{host}.googleusercontent.com/{value}"
+
+        if options and not all(param in VALID_OPTIONS for param in options.split("-")):
+            raise errors.BadRequest("Invalid options provided")
+
+        options += f"-{VALID_FMTS[file_fmt]}"
+
+        if size and size in VALID_SIZES:
+            options = f"s{size}-" + options
+
+        url += "=" + options
+
+        try:
+            response = current.request.get().response
+            response.headers["Content-Type"] = f"image/{file_fmt}"
+            response.headers["Cache-Control"] = "public, max-age=604800"  # 7 Days
+            if download:
+                response.headers["Content-Disposition"] = f"attachment; filename={filename}"
+            else:
+                response.headers["Content-Disposition"] = f"filename={filename}"
+
+            return requests.get(url).content
+
+        except Exception as e:
+            raise errors.BadRequest("Invalid Url")
```

Comment:
what kind of poor error handling is that?

```suggestion
        response = current.request.get().response
        response.headers["Content-Type"] = f"image/{file_fmt}"
        response.headers["Cache-Control"] = "public, max-age=604800"  # 7 Days
        if download:
            response.headers["Content-Disposition"] = f"attachment; filename={filename}"
        else:
            response.headers["Content-Disposition"] = f"filename={filename}"
        
        answ = requests.get(url, timeout=20)
        if not answ.ok:
            raise errors.BadRequest("Unable to fetch a file with these parameters")
```

**Notes:**
- only as much as necessary into the `try`
- Limit the requests since we have limited frontend request